v2/sourcedb-to-spanner/terraform/samples/sharded-bulk-migration/README.md

```diff
@@ -0,0 +1,418 @@
+2. Edit the `terraform.tfvars` file and replace the dummy variables with real values. Extend the configuration to meet
+   your needs. It is recommended to get started with `terraform_simple.tfvars`.
+3. Run the following commands -
+
+### Initialise Terraform
+
+```shell
+# Initialise terraform - You only need to do this once for a directory.
+terraform init
+```
+
+### Run `plan` and `apply`
+
+Validate the terraform files with -
+
+```shell
+terraform plan --var-file=terraform_simple.tfvars
+```
+
+Run the terraform script with -
+
+```shell
+terraform apply --var-file=terraform_simple.tfvars
+```
+
+This will launch the configured jobs and produce an output like below -
+
+```shell
+Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
+
+Outputs:
+
+dataflow_job_ids = [
+  "2024-06-05_00_41_11-4759981257849547781",
+]
+dataflow_job_urls = [
+  "https://console.cloud.google.com/dataflow/jobs/us-central1/2024-06-05_00_41_11-4759981257849547781",
+]
+```
+
+**Note 1:** Each of the jobs will have a random name which will be migrating 1 or more physical shards.
+
+**Note 2:** Terraform, by default, creates at most 10 resources in parallel. If you want to run more than 10 dataflow
+jobs in parallel, run terraform using the `-parallelism=n` flag. However, more parallel dataflow jobs linearly increases
+load on Spanner.
+If Spanner is not sized sufficiently, all the jobs will slow down.
+
+```shell
+terraform apply -parallelism=30
+```
+
+### Cleanup
+
+Once the jobs have finished running, you can cleanup by running -
+
+```shell
+terraform destroy
+```
+
+## FAQ
+
+### Dataflow job is failing with "Timeout in polling result file"
+
+Dataflow has a 10min timeout within which the launcher VM logic should complete. There could be multiple reasons for it
+to take over 10 mins:
+
+- **Job logs not present after the log "launcher VM started":** A sign would be there are only 3-4 log statements in the
+  job logs. This likely due to private google access is not enabled for the
+  subnetwork. Please enable private google access in your network.
+- **Job logs not present after "Discovering tables for DataSource":** This means dataflow is unable to access the mysql
+  shard. Please check your network configuration and credentials.
+- If you are still getting a Timeout after this, this means the job graph construction is taking too long. You could
+  potentially have too many tables allocated in a single Dataflow job. The recommendation is to set the `batch_size`
+  such that the total number of tables in a single job does not exceed `150`. This means if a logical shard has 15
+  tables, and a physical shard has 2 logical shards, do not set the batch_size more than 5.
+
+### Job graph is not loading/Custom counters not visible on Dataflow panel
+
+For very large graphs, this can happen. The graph section would be empty and the counters won't load. But worry not, the
+migration should progress nonetheless.
```

Comment:
This will be a blind migration and may not be acceptable to users...we should give the instructions to observe the counters directly on cloud monitoring.

CC: @bharadwaj-aditya

Reply:
They are incorrect on cloud monitoring itself, on the console they do not show up at all. There is no source of truth in these cases other than manually running count watchers on spanner.

Reply:
Are they delayed or are they incorrect? I'm confused - there must be many types of counter metrics that we publish, right?